_LowLevelCode/CPP/cpp_communicator/input_parser.cpp

```diff
@@ -113,71 +133,124 @@ prhs  --  array of pointers to right hand side parameters
     }
     else if (mex_mode.compare("labSend") == 0) {
+        if (nrhs < (int)SendInputs::N_INPUT_Arguments - 1) {
+            std::stringstream err;
+            err << " labSend needs " << (int)SendInputs::N_INPUT_Arguments - 1 << " or " << (int)SendInputs::N_INPUT_Arguments <<
+                " inputs but got " << nrhs << " input parameters\n";
+            throw_error("MPI_MEX_COMMUNICATOR:invalid_argument", err.str().c_str());
+        }
+        data_addresses.resize(1);
+        data_tag.resize(1);
+
         work_mode = labSend;
+        // the target destination address
+        data_addresses[0] = (int32_t)retrieve_value<mxInt32>("labSend: destination address", prhs[(int)SendInputs::dest_id]) - 1;
+        // the sending data tag
+        data_tag[0] = (int32_t)retrieve_value<mxInt32>("labSend: destination tag", prhs[(int)SendInputs::tag]);
+        // if the transfer is synchroneous or not
+        is_synchroneous = (bool)retrieve_value<mxUint8>("labSend: is synchronous", prhs[(int)SendInputs::is_synchronous]);
+        // retrieve pointer to serialized data to transfer
+        size_t vector_size, bytesize;
+
+        data_buffer = retrieve_vector<uint8_t >("labSend: data", prhs[(int)SendInputs::head_data_buffer], vector_size, bytesize);
+        nbytes_to_transfer = size_t(vector_size) * bytesize;
     }
     else if (mex_mode.compare("labIndex") == 0) {
         work_mode = labIndex;
     }
     else if (mex_mode.compare("labProbe") == 0) {
+        size_t n_addresses, n_tags, block_size;
+        // the queried  address
+        auto pData_addresses = retrieve_vector<mxInt32>("labProbe: source address", prhs[(int)ProbeInputs::source_id], n_addresses, block_size);
+        data_addresses.resize(n_addresses);
+        for (size_t i = 0; i < n_addresses; i++) {
+            data_addresses[i] = pData_addresses[i] - 1; // Matlab lab_index = MPI_index + 1
+        }
+        // the queried tag
+        auto pData_tag = retrieve_vector<mxInt32>("labProbe: requested tag", prhs[(int)ProbeInputs::tag], n_tags, block_size);
+        data_tag.resize(n_tags);
+        for (size_t i = 0; i < n_tags; i++) {
+            data_tag[i] = pData_tag[i];
+        }
+
         work_mode = labProbe;
     }
     else if (mex_mode.compare("barrier") == 0) {
         work_mode = labBarrier;
     }
     else if (mex_mode.compare("init") == 0) {
-        if (nrhs != 1 ) {
-            std::stringstream err;
-            err << " Init mode needs only 1 but got : " << nrhs << " input parameters";
-            throw_error("MPI_MEX_COMMUNICATOR:invalid_argument", err.str().c_str());
-        }
-        class_handle<MPI_wrapper> *pCommunicator = new class_handle<MPI_wrapper>();
         work_mode = init_mpi;
-        return pCommunicator;
+        return process_init_mode("Init", prhs, nrhs, AddPar);
+    }
+    else if (mex_mode.compare("init_test_mode") == 0) {
```

Comment:
why is `mex_mode` a string not an enum given we're switching on it?

Reply:
because enums in Matlab look horrible. And strings are pretty clear